### PR TITLE
Adds a vector store for Azure Cosmos DB for NoSQL

### PR DESCRIPTION
**Pull request by aayush3011**
This PR add supports for Azure Cosmos DB for NoSQL vector store.

Summary:

Description: added vector store integration for Azure Cosmos DB for NoSQL Vector Store,
Dependencies: azure-cosmos dependency,
Tag maintainer: @hwchase17, @baskaryan @efriis @eyurtsev

### CODE REVIEW

**Comment by aayush3011** [comment]
Any updates? @hwchase17, @baskaryan @efriis @eyurtsev

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+from azure.cosmos.database import DatabaseProxy
+from langchain_core.documents import Document
+from langchain_core.embeddings import Embeddings
+from langchain_core.vectorstores import VST, VectorStore
+
+from langchain_community.vectorstores.utils import maximal_marginal_relevance
+
+
+# You can read more about vector search using AzureCosmosDBNoSQL here.
+# https://aka.ms/CosmosVectorSearch
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:16]: just redirects to https://www.microsoft.com/en-us/ for me

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
+import uuid
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Type
+
+import numpy as np
+from azure.cosmos.container import ContainerProxy
+from azure.cosmos.cosmos_client import CosmosClient
+from azure.cosmos.database import DatabaseProxy
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:7]: should be conditional imports

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    database: DatabaseProxy
+    container: ContainerProxy
+    database_name: str
+    container_name: str
+    partition_key: str
+    embedding: Embeddings
+    vector_embedding_policy: Dict[str, Any]
+    indexing_policy: Dict[str, Any]
+    cosmos_container_properties: Dict[str, Any]
+    embedding_key: str
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:28]: VectorStore isn't a pydantic class, so currently these are class attributes (not instance attributes)

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py (hunk 1 above)
Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:16]: It should work now, as it is has now been announced publicly at [Microsoft Buiild](https://register.build.microsoft.com/ )

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py (hunk 2 above)
Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py]: Updated

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py (hunk 3 above)
Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py]: Updated

**Comment by aayush3011** [comment]
@baskaryan can you please look at the builds. I'm not sure why exactly they are failing. The PR is ready from my side.

**Comment by baskaryan** [comment]
could we add a docs page for the new integration in `docs/docs/integrations/vectorstores/`?

**Comment by aayush3011** [comment]
@baskaryan pushed the doc changes as well. Please review and merge the PR.

**Review by baskaryan — APPROVED** [review]

File: docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb
Comment (baskaryan) [inline docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb:1]: shoudl title specify 'No SQL' since there's already an Azure Cosmos DB page?

**Comment by aayush3011** [comment]
@baskaryan Can we go ahead and merge the PR.

**Comment by aayush3011** [comment]
@baskaryan Any update on this? Since we have announced Azure Cosmos DB NoSQL Vector Search on Microsoft Build Conference a couple of weeks back, a lot of customers are asking for this integration. Can we please go ahead and merge this.

**Review by baskaryan — APPROVED** [review]

File: docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb
Comment (baskaryan) [reply docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb:1]: this is what current sidebar looks like, which is confusing

Screenshot 2024-06-05 at 3 49 30 PM

Comment (aayush3011) [reply docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb:1]: What sidebar is this?

Comment (baskaryan) [reply docs/docs/integrations/vectorstores/azure_cosmos_db_no_sql.ipynb:1]: https://langchain-git-fork-aayush3011-users-akatariaco-836408-langchain.vercel.app/v0.2/docs/integrations/vectorstores/

sidebar is based on the titles of the notebooks, which is why making one of them something other than just "Azure Cosmos DB" would be helpful

**Comment by aayush3011** [comment]
@baskaryan I have updated the naming for AzureCosmosDBVectorSearch to AzureCosmosDBMongoVCoreVectorSearch. I know it might be a breaking change for some customers, but it is needed as there would be confusion when using Mongo vCore or NoSql CosmosDB. 

I tried updating the names in the langchain/langchain/vectorstore folder and tests were failing because it is deprecated to change imports in that folder. But If I don't update the names still the tests are failing. Can you please look into it, and review the PR again.

**Comment by SamJB123** [comment]
@aayush3011 Used this pull request to create a Cosmos NoSQL hosting system for [OpenGPTs](https://github.com/langchain-ai/opengpts) and it works *great!*.

One suggested addition from that experience - it could be helpful to add optional parameters for offer_throughput for both the database and the container.

Having those parameters available allows anyone to establish a whole OpenGPTs back-end instantly from an empty Cosmos NoSQL account, within the current 1000 RU/s free provisioned offering. 

(They were needed to set the database's throughput to 600 RU/s that is shared across all non-vector store containers, and also set vector store container's throughput to the minimum 400 RU/s (a necessary step given shared throughput is disabled for vector-enabled containers).

Here's the updated __init__ section with those additions:

`def __init__(
        self,
        *,
        cosmos_client: CosmosClient,
        embedding: Embeddings,
        vector_embedding_policy: Dict[str, Any],
        indexing_policy: Dict[str, Any],
        cosmos_container_properties: Dict[str, Any],
        database_name: str = "vectorSearchDB",
        container_name: str = "vectorSearchContainer",
        container_provisioned_throughput: Optional[int] = None,
        database_provisioned_throughput: Optional[int] = None,
    ):
        """
        Constructor for AzureCosmosDBNoSqlVectorSearch

        Args:
            cosmos_client: Client used to connect to azure cosmosdb no sql account.
            database_name: Name of the database to be created.
            container_name: Name of the container to be created.
            embedding: Text embedding model to use.
            vector_embedding_policy: Vector Embedding Policy for the container.
            indexing_policy: Indexing Policy for the container.
            cosmos_container_properties: Container Properties for the container.
        """
        if (
            indexing_policy["vectorIndexes"] is None
            or len(indexing_policy["vectorIndexes"]) == 0
        ):
            raise ValueError(
                "vectorIndexes cannot be null or empty in the indexing_policy."
            )
        if (
            vector_embedding_policy is None
            or len(vector_embedding_policy["vectorEmbeddings"]) == 0
        ):
            raise ValueError(
                "vectorEmbeddings cannot be null "
                "or empty in the vector_embedding_policy."
            )

        self._cosmos_client = cosmos_client
        self._database_name = database_name
        self._container_name = container_name
        self._embedding = embedding
        self._vector_embedding_policy = vector_embedding_policy
        self._indexing_policy = indexing_policy
        self._cosmos_container_properties = cosmos_container_properties
        self._container_provisioned_throughput = container_provisioned_throughput
        self._database_provisioned_throughput = database_provisioned_throughput

        # Create the database if it already doesn't exist
        self._database = self._cosmos_client.create_database_if_not_exists(
            id=self._database_name,
            offer_throughput=self._database_provisioned_throughput,
        )

        # Create the collection if it already doesn't exist
        self._container = self._database.create_container_if_not_exists(
            id=self._container_name,
            partition_key=self._cosmos_container_properties["partition_key"],
            indexing_policy=self._indexing_policy,
            vector_embedding_policy=self._vector_embedding_policy,
            offer_throughput=self._container_provisioned_throughput,
        )
        self._embedding_key = self._vector_embedding_policy["vectorEmbeddings"][0][
            "path"
        ][1:]`

**Comment by aayush3011** [comment]
@SamJB123 The container_provisioned_throughput can be provided in the cosmos_container_properties. For database_provisioned_throughput I will add a new dict as cosmos_database_properties which can container the throughput values, and other properties that can be provided when creating a database.

**Comment by aayush3011** [comment]
@baskaryan Any updates on this PR, can we get these changes in soon, a lot customers have been asking for this.

**Review by baskaryan — APPROVED** [review]
Undid the renames of the existing classes as these are breaking changes. otherwise looks good to me! can land if everything still looks good to you @aayush3011

**Comment by aayush3011** [comment]
@baskaryan we can go ahead and merge this PR. Thanks for all your help :)

**Pull request merged**